<!DOCTYPE html>

<html lang="en">

<head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=no">
    <meta name="apple-mobile-web-app-capable" content="yes">
    <meta name="apple-mobile-web-app-status-bar-style" content="black">
    <meta name="mobile-web-app-capable" content="yes">
    <title>
        8. Functional classification - HackMD
    </title>
    <link rel="icon" type="image/png" href="https://hackmd.io/favicon.png">
    <link rel="apple-touch-icon" href="https://hackmd.io/apple-touch-icon.png">

    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/ionicons/2.0.1/css/ionicons.min.css" integrity="sha256-3iu9jgsy9TpTwXKb7bNQzqWekRX7pPK+2OLj3R922fo=" crossorigin="anonymous" />
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/octicons/3.5.0/octicons.min.css" integrity="sha256-QiWfLIsCT02Sdwkogf6YMiQlj4NE84MKkzEMkZnMGdg=" crossorigin="anonymous" />
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/prism/1.5.1/themes/prism.min.css" integrity="sha256-vtR0hSWRc3Tb26iuN2oZHt3KRUomwTufNIf5/4oeCyg=" crossorigin="anonymous" />
    <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@hackmd/emojify.js@2.1.0/dist/css/basic/emojify.min.css" integrity="sha256-UOrvMOsSDSrW6szVLe8ZDZezBxh5IoIfgTwdNDgTjiU=" crossorigin="anonymous" />
    <style>
        @import url(https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,500,500i|Source+Code+Pro:300,400,500|Source+Sans+Pro:300,300i,400,400i,600,600i|Source+Serif+Pro&subset=latin-ext);.hljs{display:block;background:#fff;padding:.5em;color:#333;overflow-x:auto}.hljs-comment,.hljs-meta{color:#969896}.hljs-emphasis,.hljs-quote,.hljs-string,.hljs-strong,.hljs-template-variable,.hljs-variable{color:#df5000}.hljs-keyword,.hljs-selector-tag,.hljs-type{color:#a71d5d}.hljs-attribute,.hljs-bullet,.hljs-literal,.hljs-number,.hljs-symbol{color:#0086b3}.hljs-built_in,.hljs-builtin-name{color:#005cc5}.hljs-name,.hljs-section{color:#63a35c}.hljs-tag{color:#333}.hljs-attr,.hljs-selector-attr,.hljs-selector-class,.hljs-selector-id,.hljs-selector-pseudo,.hljs-title{color:#795da3}.hljs-addition{color:#55a532;background-color:#eaffea}.hljs-deletion{color:#bd2c00;background-color:#ffecec}.hljs-link{text-decoration:underline}.markdown-body{font-size:16px;line-height:1.5;word-wrap:break-word}.markdown-body:after,.markdown-body:before{display:table;content:""}.markdown-body:after{clear:both}.markdown-body>:first-child{margin-top:0!important}.markdown-body>:last-child{margin-bottom:0!important}.markdown-body a:not([href]){color:inherit;text-decoration:none}.markdown-body .absent{color:#c00}.markdown-body .anchor{float:left;padding-right:4px;margin-left:-20px;line-height:1}.markdown-body .anchor:focus{outline:none}.markdown-body blockquote,.markdown-body dl,.markdown-body ol,.markdown-body p,.markdown-body pre,.markdown-body table,.markdown-body ul{margin-top:0;margin-bottom:16px}.markdown-body hr{height:.25em;padding:0;margin:24px 0;background-color:#e7e7e7;border:0}.markdown-body blockquote{font-size:16px;padding:0 1em;color:#777;border-left:.25em solid #ddd}.markdown-body blockquote>:first-child{margin-top:0}.markdown-body blockquote>:last-child{margin-bottom:0}.markdown-body kbd,.popover kbd{display:inline-block;padding:3px 5px;font-size:11px;line-height:10px;color:#555;vertical-align:middle;background-color:#fcfcfc;border:1px solid #ccc;border-bottom-color:#bbb;border-radius:3px;box-shadow:inset 0 -1px 0 #bbb}.markdown-body .loweralpha{list-style-type:lower-alpha}.markdown-body h1,.markdown-body h2,.markdown-body h3,.markdown-body h4,.markdown-body h5,.markdown-body h6{margin-top:24px;margin-bottom:16px;font-weight:600;line-height:1.25}.markdown-body h1 .octicon-link,.markdown-body h2 .octicon-link,.markdown-body h3 .octicon-link,.markdown-body h4 .octicon-link,.markdown-body h5 .octicon-link,.markdown-body h6 .octicon-link{color:#000;vertical-align:middle;visibility:hidden}.markdown-body h1:hover .anchor,.markdown-body h2:hover .anchor,.markdown-body h3:hover .anchor,.markdown-body h4:hover .anchor,.markdown-body h5:hover .anchor,.markdown-body h6:hover .anchor{text-decoration:none}.markdown-body h1:hover .anchor .octicon-link,.markdown-body h2:hover .anchor .octicon-link,.markdown-body h3:hover .anchor .octicon-link,.markdown-body h4:hover .anchor .octicon-link,.markdown-body h5:hover .anchor .octicon-link,.markdown-body h6:hover .anchor .octicon-link{visibility:visible}.markdown-body h1 code,.markdown-body h1 tt,.markdown-body h2 code,.markdown-body h2 tt,.markdown-body h3 code,.markdown-body h3 tt,.markdown-body h4 code,.markdown-body h4 tt,.markdown-body h5 code,.markdown-body h5 tt,.markdown-body h6 code,.markdown-body h6 tt{font-size:inherit}.markdown-body h1{font-size:2em}.markdown-body h1,.markdown-body h2{padding-bottom:.3em;border-bottom:1px solid #eee}.markdown-body h2{font-size:1.5em}.markdown-body h3{font-size:1.25em}.markdown-body h4{font-size:1em}.markdown-body h5{font-size:.875em}.markdown-body h6{font-size:.85em;color:#777}.markdown-body ol,.markdown-body ul{padding-left:2em}.markdown-body ol.no-list,.markdown-body ul.no-list{padding:0;list-style-type:none}.markdown-body ol ol,.markdown-body ol ul,.markdown-body ul ol,.markdown-body ul ul{margin-top:0;margin-bottom:0}.markdown-body li>p{margin-top:16px}.markdown-body li+li{padding-top:.25em}.markdown-body dl{padding:0}.markdown-body dl dt{padding:0;margin-top:16px;font-size:1em;font-style:italic;font-weight:700}.markdown-body dl dd{padding:0 16px;margin-bottom:16px}.markdown-body table{display:block;width:100%;overflow:auto;word-break:normal;word-break:keep-all}.markdown-body table th{font-weight:700}.markdown-body table td,.markdown-body table th{padding:6px 13px;border:1px solid #ddd}.markdown-body table tr{background-color:#fff;border-top:1px solid #ccc}.markdown-body table tr:nth-child(2n){background-color:#f8f8f8}.markdown-body img{max-width:100%;box-sizing:content-box;background-color:#fff}.markdown-body img[align=right]{padding-left:20px}.markdown-body img[align=left]{padding-right:20px}.markdown-body .emoji{max-width:none;vertical-align:text-top;background-color:transparent}.markdown-body span.frame{display:block;overflow:hidden}.markdown-body span.frame>span{display:block;float:left;width:auto;padding:7px;margin:13px 0 0;overflow:hidden;border:1px solid #ddd}.markdown-body span.frame span img{display:block;float:left}.markdown-body span.frame span span{display:block;padding:5px 0 0;clear:both;color:#333}.markdown-body span.align-center{display:block;overflow:hidden;clear:both}.markdown-body span.align-center>span{display:block;margin:13px auto 0;overflow:hidden;text-align:center}.markdown-body span.align-center span img{margin:0 auto;text-align:center}.markdown-body span.align-right{display:block;overflow:hidden;clear:both}.markdown-body span.align-right>span{display:block;margin:13px 0 0;overflow:hidden;text-align:right}.markdown-body span.align-right span img{margin:0;text-align:right}.markdown-body span.float-left{display:block;float:left;margin-right:13px;overflow:hidden}.markdown-body span.float-left span{margin:13px 0 0}.markdown-body span.float-right{display:block;float:right;margin-left:13px;overflow:hidden}.markdown-body span.float-right>span{display:block;margin:13px auto 0;overflow:hidden;text-align:right}.markdown-body code,.markdown-body tt{padding:0;padding-top:.2em;padding-bottom:.2em;margin:0;font-size:85%;background-color:rgba(0,0,0,.04);border-radius:3px}.markdown-body code:after,.markdown-body code:before,.markdown-body tt:after,.markdown-body tt:before{letter-spacing:-.2em;content:"\00a0"}.markdown-body code br,.markdown-body tt br{display:none}.markdown-body del code{text-decoration:inherit}.markdown-body pre{word-wrap:normal}.markdown-body pre>code{padding:0;margin:0;font-size:100%;word-break:normal;white-space:pre;background:transparent;border:0}.markdown-body .highlight{margin-bottom:16px}.markdown-body .highlight pre{margin-bottom:0;word-break:normal}.markdown-body .highlight pre,.markdown-body pre{padding:16px;overflow:auto;font-size:85%;line-height:1.45;background-color:#f7f7f7;border-radius:3px}.markdown-body pre code,.markdown-body pre tt{display:inline;max-width:auto;padding:0;margin:0;overflow:visible;line-height:inherit;word-wrap:normal;background-color:transparent;border:0}.markdown-body pre code:after,.markdown-body pre code:before,.markdown-body pre tt:after,.markdown-body pre tt:before{content:normal}.markdown-body .csv-data td,.markdown-body .csv-data th{padding:5px;overflow:hidden;font-size:12px;line-height:1;text-align:left;white-space:nowrap}.markdown-body .csv-data .blob-line-num{padding:10px 8px 9px;text-align:right;background:#fff;border:0}.markdown-body .csv-data tr{border-top:0}.markdown-body .csv-data th{font-weight:700;background:#f8f8f8;border-top:0}.news .alert .markdown-body blockquote{padding:0 0 0 40px;border:0 none}.activity-tab .news .alert .commits,.activity-tab .news .markdown-body blockquote{padding-left:0}.task-list-item{list-style-type:none}.task-list-item label{font-weight:400}.task-list-item.enabled label{cursor:pointer}.task-list-item+.task-list-item{margin-top:3px}.task-list-item-checkbox{float:left;margin:.31em 0 .2em -1.3em!important;vertical-align:middle;cursor:default!important}.markdown-body{padding-top:40px;padding-bottom:40px;max-width:758px;overflow:visible!important;position:relative}.markdown-body .emoji{vertical-align:top}.markdown-body pre{border:inherit!important}.markdown-body code{color:inherit!important}.markdown-body pre code .wrapper{display:-moz-inline-flex;display:-ms-inline-flex;display:-o-inline-flex;display:inline-flex}.markdown-body pre code .gutter{float:left;overflow:hidden;-webkit-user-select:none;user-select:none}.markdown-body pre code .gutter.linenumber{text-align:right;position:relative;display:inline-block;cursor:default;z-index:4;padding:0 8px 0 0;min-width:20px;box-sizing:content-box;color:#afafaf!important;border-right:3px solid #6ce26c!important}.markdown-body pre code .gutter.linenumber>span:before{content:attr(data-linenumber)}.markdown-body pre code .code{float:left;margin:0 0 0 16px}.markdown-body .gist .line-numbers{border-left:none;border-top:none;border-bottom:none}.markdown-body .gist .line-data{border:none}.markdown-body .gist table{border-spacing:0;border-collapse:inherit!important}.markdown-body code[data-gist-id]{background:none;padding:0}.markdown-body code[data-gist-id]:after,.markdown-body code[data-gist-id]:before{content:""}.markdown-body code[data-gist-id] .blob-num{border:unset}.markdown-body code[data-gist-id] table{overflow:unset;margin-bottom:unset}.markdown-body code[data-gist-id] table tr{background:unset}.markdown-body[dir=rtl] pre{direction:ltr}.markdown-body[dir=rtl] code{direction:ltr;unicode-bidi:embed}.markdown-body .alert>p{margin-bottom:0}.markdown-body pre.abc,.markdown-body pre.flow-chart,.markdown-body pre.graphviz,.markdown-body pre.mermaid,.markdown-body pre.sequence-diagram,.markdown-body pre.vega{text-align:center;background-color:inherit;border-radius:0;white-space:inherit;overflow:visible}.markdown-body pre.abc>code,.markdown-body pre.flow-chart>code,.markdown-body pre.graphviz>code,.markdown-body pre.mermaid>code,.markdown-body pre.sequence-diagram>code,.markdown-body pre.vega>code{text-align:left}.markdown-body pre.abc>svg,.markdown-body pre.flow-chart>svg,.markdown-body pre.graphviz>svg,.markdown-body pre.mermaid>svg,.markdown-body pre.sequence-diagram>svg,.markdown-body pre.vega>svg{max-width:100%;height:100%}.markdown-body pre>code.wrap{white-space:pre-wrap;white-space:-moz-pre-wrap;white-space:-pre-wrap;white-space:-o-pre-wrap;word-wrap:break-word}.markdown-body .alert>p,.markdown-body .alert>ul{margin-bottom:0}.markdown-body summary{display:list-item}.markdown-body summary:focus{outline:none}.markdown-body details summary{cursor:pointer}.markdown-body details:not([open])>:not(summary){display:none}.markdown-body figure{margin:1em 40px}.markdown-body .mark,.markdown-body mark{background-color:#fff1a7}.vimeo,.youtube{cursor:pointer;display:table;text-align:center;background-position:50%;background-repeat:no-repeat;background-size:contain;background-color:#000;overflow:hidden}.vimeo,.youtube{position:relative;width:100%}.youtube{padding-bottom:56.25%}.vimeo img{width:100%;object-fit:contain;z-index:0}.youtube img{object-fit:cover;z-index:0}.vimeo iframe,.youtube iframe,.youtube img{width:100%;height:100%;position:absolute;top:0;left:0}.vimeo iframe,.youtube iframe{vertical-align:middle;z-index:1}.vimeo .icon,.youtube .icon{position:absolute;height:auto;width:auto;top:50%;left:50%;transform:translate(-50%,-50%);color:#fff;opacity:.3;transition:opacity .2s;z-index:0}.vimeo:hover .icon,.youtube:hover .icon{opacity:.6;transition:opacity .2s}.slideshare .inner,.speakerdeck .inner{position:relative;width:100%}.slideshare .inner iframe,.speakerdeck .inner iframe{position:absolute;top:0;bottom:0;left:0;right:0;width:100%;height:100%}.figma{display:table;position:relative;width:100%;padding-bottom:56.25%}.figma iframe{position:absolute;top:0;bottom:0;left:0;right:0;width:100%;height:100%;border:1px solid #eee}.MJX_Assistive_MathML{display:none}#MathJax_Message{z-index:1000!important}.ui-infobar{position:relative;z-index:2;max-width:760px;margin:25px auto -25px;color:#777}.toc .invisable-node{list-style-type:none}.ui-toc{position:fixed;bottom:20px;z-index:998}.ui-toc.both-mode{margin-left:8px}.ui-toc.both-mode .ui-toc-label{height:40px;padding:10px 4px;border-top-left-radius:0;border-bottom-left-radius:0}.ui-toc-label{background-color:#e6e6e6;border:none;color:#868686;transition:opacity .2s}.ui-toc .open .ui-toc-label{opacity:1;color:#fff;transition:opacity .2s}.ui-toc-label:focus{opacity:.3;background-color:#ccc;color:#000}.ui-toc-label:hover{opacity:1;background-color:#ccc;transition:opacity .2s}.ui-toc-dropdown{margin-top:20px;margin-bottom:20px;padding-left:10px;padding-right:10px;max-width:45vw;width:25vw;max-height:70vh;overflow:auto;text-align:inherit}.ui-toc-dropdown>.toc{max-height:calc(70vh - 100px);overflow:auto}.ui-toc-dropdown[dir=rtl] .nav{padding-right:0;letter-spacing:.0029em}.ui-toc-dropdown a{overflow:hidden;text-overflow:ellipsis;white-space:pre}.ui-toc-dropdown .nav>li>a{display:block;padding:4px 20px;font-size:13px;font-weight:500;color:#767676}.ui-toc-dropdown .nav>li:first-child:last-child>ul,.ui-toc-dropdown .toc.expand ul{display:block}.ui-toc-dropdown .nav>li>a:focus,.ui-toc-dropdown .nav>li>a:hover{padding-left:19px;color:#000;text-decoration:none;background-color:transparent;border-left:1px solid #000}.ui-toc-dropdown[dir=rtl] .nav>li>a:focus,.ui-toc-dropdown[dir=rtl] .nav>li>a:hover{padding-right:19px;border-left:none;border-right:1px solid #000}.ui-toc-dropdown .nav>.active:focus>a,.ui-toc-dropdown .nav>.active:hover>a,.ui-toc-dropdown .nav>.active>a{padding-left:18px;font-weight:700;color:#000;background-color:transparent;border-left:2px solid #000}.ui-toc-dropdown[dir=rtl] .nav>.active:focus>a,.ui-toc-dropdown[dir=rtl] .nav>.active:hover>a,.ui-toc-dropdown[dir=rtl] .nav>.active>a{padding-right:18px;border-left:none;border-right:2px solid #000}.ui-toc-dropdown .nav .nav{display:none;padding-bottom:10px}.ui-toc-dropdown .nav>.active>ul{display:block}.ui-toc-dropdown .nav .nav>li>a{padding-top:1px;padding-bottom:1px;padding-left:30px;font-size:12px;font-weight:400}.ui-toc-dropdown[dir=rtl] .nav .nav>li>a{padding-right:30px}.ui-toc-dropdown .nav .nav>li>ul>li>a{padding-top:1px;padding-bottom:1px;padding-left:40px;font-size:12px;font-weight:400}.ui-toc-dropdown[dir=rtl] .nav .nav>li>ul>li>a{padding-right:40px}.ui-toc-dropdown .nav .nav>li>a:focus,.ui-toc-dropdown .nav .nav>li>a:hover{padding-left:29px}.ui-toc-dropdown[dir=rtl] .nav .nav>li>a:focus,.ui-toc-dropdown[dir=rtl] .nav .nav>li>a:hover{padding-right:29px}.ui-toc-dropdown .nav .nav>li>ul>li>a:focus,.ui-toc-dropdown .nav .nav>li>ul>li>a:hover{padding-left:39px}.ui-toc-dropdown[dir=rtl] .nav .nav>li>ul>li>a:focus,.ui-toc-dropdown[dir=rtl] .nav .nav>li>ul>li>a:hover{padding-right:39px}.ui-toc-dropdown .nav .nav>.active:focus>a,.ui-toc-dropdown .nav .nav>.active:hover>a,.ui-toc-dropdown .nav .nav>.active>a{padding-left:28px;font-weight:500}.ui-toc-dropdown[dir=rtl] .nav .nav>.active:focus>a,.ui-toc-dropdown[dir=rtl] .nav .nav>.active:hover>a,.ui-toc-dropdown[dir=rtl] .nav .nav>.active>a{padding-right:28px}.ui-toc-dropdown .nav .nav>.active>.nav>.active:focus>a,.ui-toc-dropdown .nav .nav>.active>.nav>.active:hover>a,.ui-toc-dropdown .nav .nav>.active>.nav>.active>a{padding-left:38px;font-weight:500}.ui-toc-dropdown[dir=rtl] .nav .nav>.active>.nav>.active:focus>a,.ui-toc-dropdown[dir=rtl] .nav .nav>.active>.nav>.active:hover>a,.ui-toc-dropdown[dir=rtl] .nav .nav>.active>.nav>.active>a{padding-right:38px}.markdown-body{font-family:-apple-system,BlinkMacSystemFont,Segoe UI,Helvetica Neue,Helvetica,Roboto,Arial,sans-serif}html[lang^=ja] .markdown-body{font-family:-apple-system,BlinkMacSystemFont,Segoe UI,Helvetica Neue,Helvetica,Roboto,Arial,Hiragino Kaku Gothic Pro,ヒラギノ角ゴ Pro W3,Osaka,Meiryo,メイリオ,MS Gothic,ＭＳ\ ゴシック,sans-serif}html[lang=zh-tw] .markdown-body{font-family:-apple-system,BlinkMacSystemFont,Segoe UI,Helvetica Neue,Helvetica,Roboto,Arial,PingFang TC,Microsoft JhengHei,微軟正黑,sans-serif}html[lang=zh-cn] .markdown-body{font-family:-apple-system,BlinkMacSystemFont,Segoe UI,Helvetica Neue,Helvetica,Roboto,Arial,PingFang SC,Microsoft YaHei,微软雅黑,sans-serif}html .markdown-body[lang^=ja]{font-family:-apple-system,BlinkMacSystemFont,Segoe UI,Helvetica Neue,Helvetica,Roboto,Arial,Hiragino Kaku Gothic Pro,ヒラギノ角ゴ Pro W3,Osaka,Meiryo,メイリオ,MS Gothic,ＭＳ\ ゴシック,sans-serif}html .markdown-body[lang=zh-tw]{font-family:-apple-system,BlinkMacSystemFont,Segoe UI,Helvetica Neue,Helvetica,Roboto,Arial,PingFang TC,Microsoft JhengHei,微軟正黑,sans-serif}html .markdown-body[lang=zh-cn]{font-family:-apple-system,BlinkMacSystemFont,Segoe UI,Helvetica Neue,Helvetica,Roboto,Arial,PingFang SC,Microsoft YaHei,微软雅黑,sans-serif}html[lang^=ja] .ui-toc-dropdown{font-family:Source Sans Pro,Helvetica,Arial,Meiryo UI,MS PGothic,ＭＳ\ Ｐゴシック,sans-serif}html[lang=zh-tw] .ui-toc-dropdown{font-family:Source Sans Pro,Helvetica,Arial,Microsoft JhengHei UI,微軟正黑UI,sans-serif}html[lang=zh-cn] .ui-toc-dropdown{font-family:Source Sans Pro,Helvetica,Arial,Microsoft YaHei UI,微软雅黑UI,sans-serif}html .ui-toc-dropdown[lang^=ja]{font-family:Source Sans Pro,Helvetica,Arial,Meiryo UI,MS PGothic,ＭＳ\ Ｐゴシック,sans-serif}html .ui-toc-dropdown[lang=zh-tw]{font-family:Source Sans Pro,Helvetica,Arial,Microsoft JhengHei UI,微軟正黑UI,sans-serif}html .ui-toc-dropdown[lang=zh-cn]{font-family:Source Sans Pro,Helvetica,Arial,Microsoft YaHei UI,微软雅黑UI,sans-serif}.ui-affix-toc{position:fixed;top:0;max-width:15vw;max-height:70vh;overflow:auto}.back-to-top,.expand-toggle,.go-to-bottom{display:block;padding:4px 10px;margin-top:10px;margin-left:10px;font-size:12px;font-weight:500;color:#999}.back-to-top:focus,.back-to-top:hover,.expand-toggle:focus,.expand-toggle:hover,.go-to-bottom:focus,.go-to-bottom:hover{color:#563d7c;text-decoration:none}.back-to-top,.go-to-bottom{margin-top:0}.ui-user-icon{width:20px;height:20px;display:block;border-radius:50%;margin-top:2px;margin-bottom:2px;margin-right:5px;background-position:50%;background-repeat:no-repeat;background-size:cover}.ui-user-icon.small{width:18px;height:18px;display:inline-block;vertical-align:middle;margin:0 0 .2em}.ui-infobar>small>span{line-height:22px}.ui-infobar>small .dropdown{display:inline-block}.ui-infobar>small .dropdown a:focus,.ui-infobar>small .dropdown a:hover{text-decoration:none}.ui-more-info{color:#888;cursor:pointer;vertical-align:middle}.ui-more-info .fa{font-size:16px}.ui-connectedGithub,.ui-published-note{color:#888}.ui-connectedGithub{line-height:23px;white-space:nowrap}.ui-connectedGithub a.file-path{color:#888;text-decoration:none;padding-left:22px}.ui-connectedGithub a.file-path:active,.ui-connectedGithub a.file-path:hover{color:#888;text-decoration:underline}.ui-connectedGithub .fa{font-size:20px}.ui-published-note .fa{font-size:20px;vertical-align:top}.unselectable{-webkit-user-select:none;-o-user-select:none;user-select:none}.selectable{-webkit-user-select:text;-o-user-select:text;user-select:text}@media print{blockquote,div,img,pre,table{page-break-inside:avoid!important}a[href]:after{font-size:12px!important}}.markdown-body.slides{position:relative;z-index:1;color:#222}.markdown-body.slides:before{content:"";display:block;position:absolute;top:0;left:0;right:0;bottom:0;z-index:-1;background-color:currentColor;box-shadow:0 0 0 50vw}.markdown-body.slides section[data-markdown]{position:relative;margin-bottom:1.5em;background-color:#fff;text-align:center}.markdown-body.slides section[data-markdown] code{text-align:left}.markdown-body.slides section[data-markdown]:before{content:"";display:block;padding-bottom:56.23%}.markdown-body.slides section[data-markdown]>div:first-child{position:absolute;top:50%;left:1em;right:1em;transform:translateY(-50%);max-height:100%;overflow:hidden}.markdown-body.slides section[data-markdown]>ul{display:inline-block}.markdown-body.slides>section>section+section:after{content:"";position:absolute;top:-1.5em;right:1em;height:1.5em;border:3px solid #777}.site-ui-font{font-family:Source Sans Pro,Helvetica,Arial,sans-serif}html[lang^=ja] .site-ui-font{font-family:Source Sans Pro,Helvetica,Arial,Hiragino Kaku Gothic Pro,ヒラギノ角ゴ Pro W3,Osaka,Meiryo,メイリオ,MS Gothic,ＭＳ\ ゴシック,sans-serif}html[lang=zh-tw] .site-ui-font{font-family:Source Sans Pro,Helvetica,Arial,PingFang TC,Microsoft JhengHei,微軟正黑,sans-serif}html[lang=zh-cn] .site-ui-font{font-family:Source Sans Pro,Helvetica,Arial,PingFang SC,Microsoft YaHei,微软雅黑,sans-serif}body{font-smoothing:subpixel-antialiased!important;-webkit-font-smoothing:subpixel-antialiased!important;-moz-osx-font-smoothing:auto!important;text-shadow:0 0 1em transparent,1px 1px 1.2px rgba(0,0,0,.004);-webkit-overflow-scrolling:touch;letter-spacing:.025em;font-family:Source Sans Pro,Helvetica,Arial,sans-serif}html[lang^=ja] body{font-family:Source Sans Pro,Helvetica,Arial,Hiragino Kaku Gothic Pro,ヒラギノ角ゴ Pro W3,Osaka,Meiryo,メイリオ,MS Gothic,ＭＳ\ ゴシック,sans-serif}html[lang=zh-tw] body{font-family:Source Sans Pro,Helvetica,Arial,PingFang TC,Microsoft JhengHei,微軟正黑,sans-serif}html[lang=zh-cn] body{font-family:Source Sans Pro,Helvetica,Arial,PingFang SC,Microsoft YaHei,微软雅黑,sans-serif}abbr[title]{border-bottom:none;text-decoration:underline;-webkit-text-decoration:underline dotted;text-decoration:underline dotted}abbr[data-original-title],abbr[title]{cursor:help}body.modal-open{overflow-y:auto;padding-right:0!important}
    </style>
    <!-- HTML5 shim and Respond.js for IE8 support of HTML5 elements and media queries -->
    <!-- WARNING: Respond.js doesn't work if you view the page via file:// -->
    <!--[if lt IE 9]>
    	<script src="https://cdnjs.cloudflare.com/ajax/libs/html5shiv/3.7.3/html5shiv.min.js" integrity="sha256-3Jy/GbSLrg0o9y5Z5n1uw0qxZECH7C6OQpVBgNFYa0g=" crossorigin="anonymous"></script>
    	<script src="https://cdnjs.cloudflare.com/ajax/libs/respond.js/1.4.2/respond.min.js" integrity="sha256-g6iAfvZp+nDQ2TdTR/VVKJf3bGro4ub5fvWSWVRi2NE=" crossorigin="anonymous"></script>
		<script src="https://cdnjs.cloudflare.com/ajax/libs/es5-shim/4.5.9/es5-shim.min.js" integrity="sha256-8E4Is26QH0bD52WoQpcB+R/tcWQtpzlCojrybUd7Mxo=" crossorigin="anonymous"></script>
    <![endif]-->
</head>

<body>
    <div id="doc" class="markdown-body container-fluid comment-enabled" data-hard-breaks="true"><center><img src="https://i.imgur.com/sCqXiVG.png" alt="drawing" width="300"></center><h1 id="8-Functional-classification" data-id="8-Functional-classification"><a class="anchor hidden-xs" href="#8-Functional-classification" title="8-Functional-classification"><span class="octicon octicon-link"></span></a><span>8. Functional classification</span></h1><div class="alert alert-success">
<p><span>When doing a functional classification, we want to find out what the organisms in the sample are functional capable of doing. In order to do that, we need to predict the functions of the genes or reads in the sample. We do this by comparing genes or reads against databases containing sequence data with known functions.</span></p>
<p><span>In this assignment you will start with the preprocessed data from the Taxonomic classification exercise, and compare the reads against a database with functional characterised sequences. You will generate an overview of the functional potential of the species in the sample.</span></p>
<p><span>The aim of this assignment is to get familiarised with the essential steps required to perform function classification of metagenomic reads, and comparison of the functional profile from multiple samples</span></p>
<p><strong><span>In this exercise you will:</span></strong></p>
<ol>
<li><span>Functional annotation of reads using </span><strong><span>SUPER-FOCUS</span></strong></li>
<li><span>Visualisation of multiple functional profiles using </span><strong><span>STAMP</span></strong></li>
<li><span>Pathway analysis using </span><strong><span>HUMMan2</span></strong></li>
</ol>
<h4 id="Good-luck-with-the-exercises-and-have-fun👍😃👍" data-id="Good-luck-with-the-exercises-and-have-fun👍😃👍"><a class="anchor hidden-xs" href="#Good-luck-with-the-exercises-and-have-fun👍😃👍" title="Good-luck-with-the-exercises-and-have-fun👍😃👍"><span class="octicon octicon-link"></span></a><span>Good luck with the exercises and have fun👍😃👍</span></h4>
</div><h2 id="1-Functional-annotation-of-reads" data-id="1-Functional-annotation-of-reads"><a class="anchor hidden-xs" href="#1-Functional-annotation-of-reads" title="1-Functional-annotation-of-reads"><span class="octicon octicon-link"></span></a><span>1. Functional annotation of reads</span></h2><center><img src="https://i.imgur.com/fdkadG1.png" alt="drawing" width="400"></center><br><div class="alert alert-success">
<p><span>💡</span><strong><span>SUPER-FOCUS</span></strong><span>, SUbsystems Profile by databasE Reduction using FOCUS, an agile homology-based approach using a reduced reference database to report the subsystems present in metagenomic datasets and profile their abundances. Read more </span><a href="https://github.com/metageni/SUPER-FOCUS" target="_blank" rel="noopener"><span>here</span></a></p>
</div><div class="alert alert-danger">
<p><span>Running </span><strong><span>SUPER-FOCUS</span></strong><span> for a single sample will take 2-3 hours. Therefore, we have prerun </span><strong><span>SUPER-FOCUS</span></strong><span> for you. The tool is installed, so if time allows, feel free to test the tool for yourself.</span></p>
<h4 id="This-is-how-SUPER-FOCUS-was-run-on-Sample-1" data-id="This-is-how-SUPER-FOCUS-was-run-on-Sample-1"><a class="anchor hidden-xs" href="#This-is-how-SUPER-FOCUS-was-run-on-Sample-1" title="This-is-how-SUPER-FOCUS-was-run-on-Sample-1"><span class="octicon octicon-link"></span></a><span>This is how SUPER-FOCUS was run on Sample 1:</span></h4>
<p><span style="color:blue"><span>I] Go to the directory </span><kbd><span>~/practical/</span></kbd><span> and copy the data for this exercise from the shared disk:</span></span></p>
<pre><code>cp -r /net/software/practical/8 .
</code></pre>
<div class="alert alert-info">
<p><img class="emoji" alt=":information_source:" src="https://cdn.jsdelivr.net/npm/@hackmd/emojify.js@2.1.0/dist/images/basic/information_source.png"><span> Remember to give yourself write permissions to the directory (and sub directories)</span></p>
</div>
<p><span style="color:blue"><span>In the directory </span><kbd><span>~/practical/8/superfocus/input/</span></kbd><span>, we have put a copy of the preprocessed data from Exercise 3 (</span><code>sample1_R1_pair.fastq.gz</code><span> and </span><code>sample1_R1_pair.fastq.gz</code><span>)</span></span></p>
<p><span style="color:blue"><span>The R1 and R2 FASTQ files were concatenated and named </span><code>Sample1_pair.fastq.gz</code></span></p>
<p><span style="color:blue"><span>All files in the directory </span><kbd><span>~/practical/11/superfocus/input</span></kbd><span> except </span><code>Sample1_pair.fastq.gz</code><span> were deleted</span></span></p>
<p><span style="color:blue"><span>In the directory </span><kbd><span>~/practical/8/superfocus</span></kbd><span>, </span><strong><span>SUPER-FOCUS</span></strong><span> was run on the FASTA file to functional annotate all the reads:</span></span></p>
<br>
<pre><code>superfocus -q input -dir sample1 -a diamond -n 0 -t 16

#-q, input directory
#-d, output directory
#-a, mapping tool
#-n, normalises each query counts
#-t, number of threads
</code></pre>
</div><div class="alert alert-info">
<p><img class="emoji" alt=":information_source:" src="https://cdn.jsdelivr.net/npm/@hackmd/emojify.js@2.1.0/dist/images/basic/information_source.png"><span> </span><strong><span>SUPER-FOCUS</span></strong><span> output will be add the folder selected by the -dir argument. The main output is the </span><code>output_all_levels_and_function.xls</code><span> file, but </span><strong><span>SUPER-FOCUS</span></strong><span> also produce a summary for each of the three hierarchical subsystem levels.</span></p>
<p><span>In the main output file the three subsystem levels are in column 1-3, while the gene function is in column 4. For single sample analysis, the 5th column is the number of reads assigned to this subsystem, and the 6th column the corresponding relative abundance value of this subsystem in this sample.</span></p>
<br>
<table>
<thead>
<tr>
<th><span>Subsystem Level 1</span></th>
<th><span>Subsystem Level 2</span></th>
<th><span>Subsystem Level 3</span></th>
<th><span>Function</span></th>
<th><span>Sample2_cat.fasta</span></th>
<th><span>Sample2_cat.fasta %</span></th>
</tr>
</thead>
<tbody>
<tr>
<td><span>Amino Acids and Derivatives</span></td>
<td><span>-</span></td>
<td><span>Amino acid racemase</span></td>
<td><span>2-methylaconitate_cis-trans_isomerase</span></td>
<td><span>2</span></td>
<td><span>4.027686315734358e-05</span></td>
</tr>
<tr>
<td><span>Amino Acids and Derivatives</span></td>
<td><span>-</span></td>
<td><span>Amino acid racemase</span></td>
<td><span>4-oxalomesaconate_tautomerase_(EC_5.3.2.8)</span></td>
<td><span>37</span></td>
<td><span>0.0007451219684108562</span></td>
</tr>
<tr>
<td><span>…</span></td>
<td><span>…</span></td>
<td><span>…</span></td>
<td><span>…</span></td>
<td><span>…</span></td>
<td><span>…</span></td>
</tr>
</tbody>
</table>
</div><p><span style="color:blue"><span>I] The output from </span><strong><span>SUPER-FOCUS</span></strong><span> is in the </span><code>sample1</code><span> directory. Use the summary files in the </span><code>sample1</code><span> directory to answer the following questions:</span></span></p><p><span>❓How many different subsystems level 1 has been identified in the sample?</span></p><div class="alert alert-danger">
<p><span>Fill in the answer in the shared </span><a href="https://docs.google.com/spreadsheets/d/1JQmH5j6ygClMxUfMn654lhGKU560LoSAXW61DA03Zg8/edit?usp=sharing" target="_blank" rel="noopener"><span>Google docs table</span></a></p>
</div><p><span>❓How large fraction of the reads in the sample have been mapped to the subsystem Amino Acids and Derivatives?</span></p><details>
<summary>Solution - Click to expand 
</summary>
<div style="background-color:#c9cace">
9.6%
</div></details><br><p><span>❓Which subsystems level 1 is most abundant in your sample and how large fraction of the reads maps to this?</span></p><div class="alert alert-danger">
<p><span>Fill in the answer in the shared </span><a href="https://docs.google.com/spreadsheets/d/1JQmH5j6ygClMxUfMn654lhGKU560LoSAXW61DA03Zg8/edit?usp=sharing" target="_blank" rel="noopener"><span>Google docs table</span></a></p>
</div><p><span>❓Which function is most abundant within the most abundant subsystem, and how many reads in your sample is classified with this function?</span></p><details>
<summary>Solution - Click to expand 
</summary>
<div style="background-color:#c9cace">
Beta-galactosidase_(EC_3.2.1.23), 14311 reads
</div></details><br><div class="alert alert-warning">
<h4 id="-Optional-exercise" data-id="-Optional-exercise"><a class="anchor hidden-xs" href="#-Optional-exercise" title="-Optional-exercise"><span class="octicon octicon-link"></span></a><img class="emoji" alt=":white_check_mark:" src="https://cdn.jsdelivr.net/npm/@hackmd/emojify.js@2.1.0/dist/images/basic/white_check_mark.png"><span> Optional exercise</span></h4>
<p><span style="color:blue"><span>Make a simple bar plot showing the distribution of the subsystem level 1 for this sample.</span></span></p>
<br>
<center><img src="https://i.imgur.com/xd7e3wH.png" alt="drawing" width="800"></center>
</div><summary><span style="font-weight:bold">Progress tracker</span></summary><div style="padding:1px;background:#CCC;">
 <div style="width:33%;background:#291;text-align:center;">
   <span style="color:white"><span>Part 1 finished</span>
 </span></div>
</div><h2 id="2-Visualisation-of-functional-profiles" data-id="2-Visualisation-of-functional-profiles"><a class="anchor hidden-xs" href="#2-Visualisation-of-functional-profiles" title="2-Visualisation-of-functional-profiles"><span class="octicon octicon-link"></span></a><span>2. Visualisation of functional profiles</span></h2><center><img src="https://i.imgur.com/pDrbnrV.png" alt="drawing" width="200"></center><div class="alert alert-info">
<p><span>💡</span><strong><span>STAMP</span></strong><span> is a software package for analysing taxonomic or metabolic profiles that promotes ‘best practices’ in choosing appropriate statistical techniques and reporting results. Normally you would use this tools to compare and visualise the functional profiles from multiple samples.</span></p>
<p><span>The graphical interface permits easy exploration of statistical results and generation of publication quality plots for inferring the biological relevance of features in a metagenomic profile.</span></p>
<p><span>Read more </span><a href="http://kiwi.cs.dal.ca/Software/STAMP" target="_blank" rel="noopener"><span>here</span></a></p>
<h4 id="STAMP-requires-two-input-files" data-id="STAMP-requires-two-input-files"><a class="anchor hidden-xs" href="#STAMP-requires-two-input-files" title="STAMP-requires-two-input-files"><span class="octicon octicon-link"></span></a><span>STAMP requires two input files:</span></h4>
<ol>
<li><span>Profile file - relative abundances</span></li>
<li><span>Metadata file</span></li>
</ol>
<p><span>Example of Profile file:</span></p>
<br>
<table>
<thead>
<tr>
<th><span>Subsystem Level 1</span></th>
<th><span>Sample1</span></th>
<th><span>Sample2</span></th>
<th><span>Sample3</span></th>
<th><span>…</span></th>
</tr>
</thead>
<tbody>
<tr>
<td><span>Amino Acids and Derivatives</span></td>
<td><span>7.8</span></td>
<td><span>5.6</span></td>
<td><span>4.5</span></td>
<td><span>…</span></td>
</tr>
<tr>
<td><span>Carbohydrates</span></td>
<td><span>2.6</span></td>
<td><span>4.9</span></td>
<td><span>1.5</span></td>
<td><span>…</span></td>
</tr>
</tbody>
</table>
<p><span>Example of Metadata file:</span></p>
<br>
<table>
<thead>
<tr>
<th><span>#SampleID</span></th>
<th><span>metadata1</span></th>
<th><span>metadata2</span></th>
<th><span>…</span></th>
</tr>
</thead>
<tbody>
<tr>
<td><span>Sample1</span></td>
<td><span>treated</span></td>
<td><span>0mnd</span></td>
<td><span>…</span></td>
</tr>
<tr>
<td><span>Sample2</span></td>
<td><span>control</span></td>
<td><span>6mnd</span></td>
<td><span>…</span></td>
</tr>
</tbody>
</table>
</div><div class="alert alert-danger">
<p><span>We have prerun </span><strong><span>SUPER-FOCUS</span></strong><span> on all the eight samples you worked with when you were preforming taxonomic analysis. The results are available here: </span><kbd><span>~/practical/8/superfocus_all</span></kbd></p>
</div><p><span style="color:blue"><span>I] Go tho the directory with the prerun </span><strong><span>SUPER-FOCUS</span></strong><span> results. Open the </span><code>output_subsystem_level_1.xls</code><span> and remove the columns with the counts, leaving only relative abundance values in the table. Also remove the  four top rows as this information is not relevant for the analysis. Finally, remove </span><code>_pair.fastq %</code><span> from the sample names.</span></span></p><p><span>The start of the file should look something like this:</span></p><pre><code>Subsystem 1	Sample1	Sample2	Sample3	Sample4	Sample5	Sample6	Sample7	Sample8
Amino Acids and Derivatives	9.964002469006855	9.722291028530115	9.374379234996798	10.051818807098806	9.966066877163616	11.05482606310993	11.473045070460687	9.86607248678503
Carbohydrates	17.868022243147482	17.736943751346757	19.04047203890495	11.825792048620691	11.887320728877391	14.201276088910813	13.522169145672388	13.348424870220684
</code></pre><p><span style="color:blue"><span>II] Save the new file as a tab-delimited text file and name it </span><code>subsystem_level_1.txt</code></span></p><p><span style="color:blue"><span>III] Check the metadata file (same as for the taxonomic profile exercise). Make sure the data is in the correct format and that the samples have the same names as in </span><code>subsystem_level_1.txt</code><span>.</span></span></p><p><span style="color:blue"><span>IV] Start </span><strong><span>STAMP</span></strong><span> from the </span><strong><span>Applications</span></strong><span> menu. Select </span><strong><span>Wine</span></strong><span> -&gt; </span><strong><span>Programs</span></strong><span> -&gt; </span><strong><span>STAMP</span></strong><span> -&gt; </span><strong><span>STAMP</span></strong></span></p><p><span style="color:blue"><span>V] From the </span><strong><span>“File”</span></strong><span> menu, choose </span><strong><span>“Load data”</span></strong><span>. In the dialog box that opens, browser to the directory </span><kbd><span>~/practical/8/superfocus_all</span></kbd><span> and choose </span><code>subsystem_level_1.txt</code><span> as profile file and </span><code>metadata.txt</code><span> as metadata file, and press </span><strong><span>“OK”</span></strong><span>.</span></span></p><center><img src="https://i.imgur.com/hs28wB5.png" alt="drawing" width="400"></center><br><div class="alert alert-info">
<p><img class="emoji" alt=":information_source:" src="https://cdn.jsdelivr.net/npm/@hackmd/emojify.js@2.1.0/dist/images/basic/information_source.png"><span> A new window should appear with a graphical interface that permits easy exploration of statistical results and generation of publication-quality plots for inferring the biological relevance of features in a metagenomic profile. There may be other default settings in </span><strong><span>STAMP</span></strong><span> that will cause the opening window to look different from the screen shot - but don’t worry!</span></p>
<p><span>The graphical interface is by default divided into sections with the </span><strong><span>Property</span></strong><span> section to the left, the </span><strong><span>Plot</span></strong><span> window in the middle and the </span><strong><span>Group</span></strong><span> section to the right. You can rearrange the organisation of the windows if you like. </span><strong><span>STAMP</span></strong><span> may also open with the </span><strong><span>Metadata</span></strong><span> table at the bottom view.</span></p>
<p><span>Spend some minutes to get to know this graphical interface. Don’t worry, you will not destroy your data in </span><strong><span>STAMP</span></strong><span>.</span></p>
</div><center><img src="https://i.imgur.com/4POXWGK.png" alt="drawing" width="800"></center><br><p><span style="color:blue"><span>VI] The default plot is a PCA plot of the data. Colour the plot by the metadata category </span><strong><span>Group</span></strong><span>. You may need to adjust the size of the plot window first. Also, make sure that you only display the PC1 vs PC2 plot by pressing the “Configure plot” button below the plot and deselcting the other plots (if they are displayed by default).</span></span></p><p><span>❓Does the samples from the patients treated with probiotics differ from the untreated patients, and which PCA components explain the variation?</span></p><details>
<summary>Solution - Click to expand 
</summary>
<div style="background-color:#c9cace">
Yes, clear separation. 80,3% of the variation can be explained by PCA1
</div></details><br><div class="alert alert-info">
<p><img class="emoji" alt=":information_source:" src="https://cdn.jsdelivr.net/npm/@hackmd/emojify.js@2.1.0/dist/images/basic/information_source.png"><span> Inspired by the </span><strong><span>STAMP</span></strong><span> User’s Guide: You can read this later…</span></p>
<p><span>There is no minimum sample size required for a statistical hypothesis test to be valid, but the assumptions for the test statistic must be met (e.g., approximately normally distributed). Small sample sizes are more likely to violate these assumptions. A small sample size is also less likely to have the statistical power required to identify a small effect size as statistically significant.</span></p>
<p><span>Biological data is noisy. Taxonomic and metabolic profiles are subjected to a lot of variability. Changing the method used to classify sequences or the underlying reference database will often result in substantial changes to the resulting profiles. Sample preparation will also influence the resulting profiles. Intuitively, we expect biological replicates to produce similar profiles, but we accept that there will be a lot of variability. We are also often comparing broadly defined groups where we expect the intragroup variability to be substantial, e.g., community profiles of healthy vs sick individuals. Intuitively, a large number of samples will be required to reliably estimate the mean and variance of a group under these conditions. Consequently, more samples per a group are required before it is reasonable to compare the means of these two groups. The exact number of samples required depends on the effect size between these groups, the desired alpha level for defining statistical significance, and the desired statistical power.</span></p>
<p><span>Effect sizes must also be considered when assessing results. A feature with a statistically significant difference between two groups, regardless of sample size, may not be biologically relevant. When sample sizes are large, even extremely small differences will be statistically significant. However, caution is warranted when effect sizes are small as statistical tests do not account for systematic biases that may exists in the methodology used to generate a taxonomic or metabolic profile. For example, a small increase in Firmicutes in 100 healthy patients vs. 100 sick patients may simply be the result of reference databases containing more Firmicute species found within healthy humans. When sample sizes are small, the reported p-values will often be inaccurate as statistical hypothesis test cannot account for the poor accuracy and precision of the methods used to generate taxonomic and metabolic profiles.</span></p>
</div><p><span style="color:blue"><span>VII] Test which functional categories that are significantly different between the two groups. Under the </span><strong><span>“Two groups”</span></strong><span> dialog on the left, check that </span><strong><span>“Welch’s t-test”</span></strong><span> is being used as the statistical test, and select </span><strong><span>“Benjamini-Hochberg FDR”</span></strong><span> for the multiple test correction.</span></span></p><center><img src="https://i.imgur.com/ocLaC25.png" alt="drawing" width="400"></center><br><p><span style="color:blue"><span>VIII] Press the </span><strong><span>“i”</span></strong><span> symbol next to </span><strong><span>“Benjamini-Hochberg FDR”</span></strong><span>.</span></span></p><p><span>❓How many features (functional categories) are significantly different between samples based on treatment with probiotics?</span></p><div class="alert alert-danger">
<p><span>Fill in the answer in the shared </span><a href="https://docs.google.com/spreadsheets/d/1JQmH5j6ygClMxUfMn654lhGKU560LoSAXW61DA03Zg8/edit?usp=sharing" target="_blank" rel="noopener"><span>Google docs table</span></a></p>
</div><div class="alert alert-info">
<p><img class="emoji" alt=":information_source:" src="https://cdn.jsdelivr.net/npm/@hackmd/emojify.js@2.1.0/dist/images/basic/information_source.png"><span> The Benjamini-Hochberg Procedure is a powerful tool that decreases the false discovery rate. The statistics is beyond the scope of this course, but you can read more </span><a href="https://en.wikipedia.org/wiki/False_discovery_rate" target="_blank" rel="noopener"><span>here</span></a></p>
</div><p><span style="color:blue"><span>IX] Change the multiple test correction to “</span><strong><span>No correction</span></strong><span>”, which will perform the statistical testing without the false discovery rate.</span></span></p><p><span>❓How many features (functional categories) are significantly different between samples based on treatment with probiotics?</span></p><div class="alert alert-danger">
<p><span>Fill in the answer in the shared </span><a href="https://docs.google.com/spreadsheets/d/1JQmH5j6ygClMxUfMn654lhGKU560LoSAXW61DA03Zg8/edit?usp=sharing" target="_blank" rel="noopener"><span>Google docs table</span></a></p>
</div><p><span style="color:blue"><span>X] Open a tabular view of the results from the </span><strong><span>“View”</span></strong><span> -&gt; </span><strong><span>“Two group statistics table”</span></strong><span> menu item.</span></span></p><center><img src="https://i.imgur.com/Agosmv1.png" alt="drawing" width="800"></center><br><div class="alert alert-info">
<p><img class="emoji" alt=":information_source:" src="https://cdn.jsdelivr.net/npm/@hackmd/emojify.js@2.1.0/dist/images/basic/information_source.png"><span> Columns can be sorted to help identify patterns of interest. Results can be limited to only the active features (those passing the specified filters) by checking the </span><strong><span>Show only active features</span></strong><span> checkbox. The table can be saved to file using the </span><strong><span>Save</span></strong><span> button. Tables are saved as text files in tab-separated values format which can be read by any text editor and most spreadsheet programs.</span></p>
</div><p><span>❓Which features (functional categories) are significantly different between samples based on treatment with probiotics?</span></p><details>
<summary>Solution - Click to expand 
</summary>
<div style="background-color:#c9cace">
<ul>
<li><span>Amino Acids and Derivatives</span></li>
<li><span>Carbohydrates</span></li>
<li><span>Cell Wall and Capsule</span></li>
<li><span>Cofactors, Vitamins, Prosthetic Groups, Pigments</span></li>
<li><span>DNA Metabolism</span></li>
<li><span>Miscellaneous</span></li>
<li><span>Phosphorus Metabolism</span></li>
<li><span>Protein Metabolism</span></li>
<li><span>Regulation and Cell signaling</span></li>
<li><span>Respiration</span></li>
<li><span>RNA Metabolism</span></li>
<li><span>Stress Response</span></li>
<li><span>Virulence, Disease and Defense</span></li>
</ul>
</div></details><br><p><span style="color:blue"><span>XI] Close the </span><strong><span>“Two group statistics table”</span></strong><span>.</span></span></p><p><span style="color:blue"><span>XII] Change the metadata category you are making group comparisons between.</span></span></p><p><span>❓Does the mode of birth delivery (Section/VF) seem to have any impact on the clustering of samples based on the functional profiles?</span></p><details>
<summary>Solution - Click to expand 
</summary>
<div style="background-color:#c9cace">
No. Babies born either vaginal or by C-section is not clustered apart
</div></details><br><p><span style="color:blue"><span>XIII] Change the metadata category back to </span><strong><span>Groups</span></strong><span>. Open the box plot and select </span><strong><span>Carbohydrates</span></strong><span>.</span></span></p><p><span>❓What are the mean abundance for the two groups? Hint: You can view the numbers in the </span><strong><span>“Two group statistics table”</span></strong></p><details>
<summary>Solution - Click to expand 
</summary>
<div style="background-color:#c9cace">
Treated 15.87%. Untreated 17.70%
</div></details><br><p><span>❓Are the two groups significantly different?</span></p><details>
<summary>Solution - Click to expand 
</summary>
<div style="background-color:#c9cace">
Yes, p-value 7.24e-3
</div></details><br><p><span style="color:blue"><span>XIV] Open the bar plot for for the same functional category.</span></span></p><p><span>❓Does the relative abundance data look consistent for the samples within the two groups?</span></p><details>
<summary>Solution - Click to expand 
</summary>
<div style="background-color:#c9cace">
Yes very... Relatively higher abundance for all untreated samples
</div></details><br><p><span style="color:blue"><span>XV] Generate a heatmap for the data.</span></span></p><p><span>❓Which functional category has highest fractional abundance?</span></p><div class="alert alert-danger">
<p><span>Fill in the answer in the shared </span><a href="https://docs.google.com/spreadsheets/d/1JQmH5j6ygClMxUfMn654lhGKU560LoSAXW61DA03Zg8/edit?usp=sharing" target="_blank" rel="noopener"><span>Google docs table</span></a></p>
</div><p><span>❓Does the relative abundance data look consistent for the samples within the two groups in the heatmap?</span></p><details>
<summary>Solution - Click to expand 
</summary>
<div style="background-color:#c9cace">
Yes very...
</div></details><br><div class="alert alert-info">
<p><img class="emoji" alt=":information_source:" src="https://cdn.jsdelivr.net/npm/@hackmd/emojify.js@2.1.0/dist/images/basic/information_source.png"><span> It is possible to digg deeper into each functional category by preparing the SUPER-FOCUS results from subsystem level 2 and 3. But unfortunately we do not have time to do this in this course😩.</span></p>
</div><p><span style="color:blue"><span>XV] Close the </span><strong><span>STAMP</span></strong><span> tool</span></span></p><summary><span style="font-weight:bold">Progress tracker</span></summary><div style="padding:1px;background:#CCC;">
 <div style="width:66%;background:#291;text-align:center;">
   <span style="color:white"><span>Part 2 finished</span>
 </span></div>
</div><h2 id="3-Pathway-analysis-using-HUMMan2" data-id="3-Pathway-analysis-using-HUMMan2"><a class="anchor hidden-xs" href="#3-Pathway-analysis-using-HUMMan2" title="3-Pathway-analysis-using-HUMMan2"><span class="octicon octicon-link"></span></a><span>3. Pathway analysis using HUMMan2</span></h2><center><img src="https://i.imgur.com/6imlkIx.jpg" alt="drawing" width="500"></center><br><div class="alert alert-success">
<p><span>💡</span><strong><span>HUMAnN2</span></strong><span> is a pipeline for efficiently and accurately profiling the presence/absence and abundance of microbial pathways in a community from metagenomic or metatranscriptomic sequencing data. You can read more about the tool </span><a href="https://bitbucket.org/biobakery/humann2/wiki/Home" target="_blank" rel="noopener"><span>here</span></a></p>
<p><strong><span>HUMAnN3</span></strong><span> has recently been released, but has not been tested extensively, so we decided to use </span><strong><span>HUMAnN2</span></strong></p>
</div><div class="alert alert-danger">
<p><strong><span>HUMAnN2</span></strong><span> is resource demanding and requires at least 16GB RAM to run. The runtime for a sample of the size you are working with in this course is typical  2-3 hours on a 512GB RAM machine. We have therefore prerun </span><strong><span>HUMAnN2</span></strong><span> on all the 8 metagenomic samples.</span></p>
<h4 id="This-is-how-HUMAnN2-was-run-on-Sample-1" data-id="This-is-how-HUMAnN2-was-run-on-Sample-1"><a class="anchor hidden-xs" href="#This-is-how-HUMAnN2-was-run-on-Sample-1" title="This-is-how-HUMAnN2-was-run-on-Sample-1"><span class="octicon octicon-link"></span></a><span>This is how HUMAnN2 was run on Sample 1:</span></h4>
<p><span style="color:blue"><span>Concatenate all reads into a single FASTQ file:</span></span></p>
<br>
<pre><code>cat Sample1_R1.fq.gz Sample1_R2.fq.gz &gt; Sample1_pair.fastq
</code></pre>
<p><span style="color:blue"><span>Run </span><strong><span>HUMAnN2</span></strong><span> on </span><code>Sample1_pair.fastq</code></span></p>
<br>
<pre><code>humann2 --input Sample1_pair.fastq.gz --output Sample1_humman2 --memory-use maximum

# --input, input files
# --output, directory to write output files
# --memory-use
 
OUTPUT:

Sample1_humman2/Sample1_pair_genefamilies.tsv
Sample1_humman2/Sample1_pair_pathcoverage.tsv
Sample1_humman2/Sample1_pair_pathabundance.tsv
</code></pre>
<h5 id="NB-If-you-set-this-up-on-your-own-system-you-might-have-to-change-the-path-to-the-databases-including-the-MetaPhlAn2-databases-Read-here" data-id="NB-If-you-set-this-up-on-your-own-system-you-might-have-to-change-the-path-to-the-databases-including-the-MetaPhlAn2-databases-Read-here"><a class="anchor hidden-xs" href="#NB-If-you-set-this-up-on-your-own-system-you-might-have-to-change-the-path-to-the-databases-including-the-MetaPhlAn2-databases-Read-here" title="NB-If-you-set-this-up-on-your-own-system-you-might-have-to-change-the-path-to-the-databases-including-the-MetaPhlAn2-databases-Read-here"><span class="octicon octicon-link"></span></a><span>NB! If you set this up on your own system, you might have to change the path to the databases, including the </span><strong><span>MetaPhlAn2</span></strong><span> databases. Read </span><a href="https://bitbucket.org/biobakery/humann2/wiki/Home#markdown-header-5-download-the-databases" target="_blank" rel="noopener"><span>here</span></a></h5>
<h5 id="NB-HUMAnN2---must-use-diamond-v0822" data-id="NB-HUMAnN2---must-use-diamond-v0822"><a class="anchor hidden-xs" href="#NB-HUMAnN2---must-use-diamond-v0822" title="NB-HUMAnN2---must-use-diamond-v0822"><span class="octicon octicon-link"></span></a><span>NB! HUMAnN2 - must use diamond v0.8.22</span></h5>
</div><p><span style="color:blue"><span>I] Go to the directory where the </span><strong><span>HUMAnN2</span></strong><span> outputs are located </span><kbd><span>practical/8/humann2/all</span></kbd><span> and view the content of this directory.</span></span></p><div class="alert alert-info">
<p><img class="emoji" alt=":information_source:" src="https://cdn.jsdelivr.net/npm/@hackmd/emojify.js@2.1.0/dist/images/basic/information_source.png"><span> </span><strong><span>HUMAnN2</span></strong><span> generates three main output files per sample:</span></p>
<p><strong><span>Gene families file:</span></strong></p>
<ul>
<li><span>This file details the abundance of each gene family in the community. Gene families are groups of evolutionarily-related protein-coding sequences that often perform similar functions.</span></li>
<li><span>Gene family abundance at the community level is stratified to show the contributions from known and unknown species. Individual species’ abundance contributions sum to the community total abundance.</span></li>
<li><span>HUMAnN2 uses the MetaPhlAn2 software along with the ChocoPhlAn database and translated search database for this computation.</span></li>
<li><span>Gene family abundance is reported in RPK (reads per kilobase) units to normalize for gene length; RPK units reflect relative gene (or transcript) copy number in the community. RPK values can be further sum-normalized to adjust for differences in sequencing depth across samples.</span></li>
</ul>
<br>
<p><strong><span>Pathway abundance file:</span></strong></p>
<ul>
<li><span>This file details the abundance of each pathway in the community as a function of the abundances of the pathway’s component reactions, with each reaction’s abundance computed as the sum over abundances of genes catalyzing the reaction.</span></li>
<li><span>Pathway abundance is computed once at the community level and again for each species (plus the “unclassified” stratum) using community- and species-level gene abundances along with the structure of the pathway.</span></li>
<li><span>The pathways are ordered by decreasing abundance with pathways for each species also sorted by decreasing abundance. Pathways with zero abundance are not included in the file.</span></li>
</ul>
<br>
<p><strong><span>Pathway coverage file:</span></strong></p>
<ul>
<li><span>Pathway coverage provides an alternative description of the presence (1) and absence (0) of pathways in a community, independent of their quantitative abundance.</span></li>
<li><span>More specifically, HUMAnN2 assigns a confidence score to each reaction detected in the community. Reactions with abundance greater than the median reaction abundance are considered to be more confidently detected than those below the median abundance.</span></li>
<li><span>HUMAnN2 then computes pathway coverage using the same algorithms described above in the context of pathway abundance, but substituting reaction confidence for reaction abundance.</span></li>
<li><span>A pathway with coverage = 1 is considered to be confidently detected (independent of its abundance), as this implies that all of its member reactions were also confidently detected. A pathway with coverage = 0 is considered to less confidently detected (independent of its abundance), as this implies that some of its member reactions were not confidently detected.</span></li>
<li><span>Like pathway abundance, pathway coverage is computed for the community as a whole, as well as for each detected species and the unclassified stratum.</span></li>
<li><span>Much as community-level pathway abundance is not the strict sum of species-level contributions, </span><strong><span>it is possible for a pathway to be confidently covered at the community level but never confidently detected from any single species.</span></strong></li>
</ul>
</div><p><span style="color:blue"><span>II] Explore the output files from Sample 1 and answer the following:</span></span></p><p><span>❓Except from the Unmapped and the Unknown, what is the most abundant gene family in Sample 1? Hint: Check in </span><a href="https://www.uniprot.org/" target="_blank" rel="noopener"><span>UniProt</span></a></p><div class="alert alert-danger">
<p><span>Fill in the answer in the shared </span><a href="https://docs.google.com/spreadsheets/d/1JQmH5j6ygClMxUfMn654lhGKU560LoSAXW61DA03Zg8/edit?usp=sharing" target="_blank" rel="noopener"><span>Google docs table</span></a></p>
</div><p><span>❓How much of the contribution to this gene family abundance comes from unclassified species (in RPK)?</span></p><details>
<summary>Solution - Click to expand 
</summary>
<div style="background-color:#c9cace">
10.2
</div></details><br><p><span>❓Except from the Unmapped and the Unintegrated, what is the most abundant pathway in Sample 1?</span></p><div class="alert alert-danger">
<p><span>Fill in the answer in the shared </span><a href="https://docs.google.com/spreadsheets/d/1JQmH5j6ygClMxUfMn654lhGKU560LoSAXW61DA03Zg8/edit?usp=sharing" target="_blank" rel="noopener"><span>Google docs table</span></a></p>
</div><p><span>❓A pathway with coverage = 1 is considered to be confidently detected. How confident is this pathway detected for </span><em><span>Bifidobacterium longum</span></em><span>?</span></p><details>
<summary>Solution - Click to expand 
</summary>
<div style="background-color:#c9cace">
0.9999 (1 is considered to be confidently detected)
</div></details><br><p><span>❓How confident is this pathway “ARGSYN-PWY: L-arginine biosynthesis I (via L-ornithine)” detected for </span><em><span>Bifidobacterium longum</span></em><span>?</span></p><details>
<summary>Solution - Click to expand 
</summary>
<div style="background-color:#c9cace">
0, less confidently detected
</div></details><br><div class="alert alert-info">
<p><img class="emoji" alt=":information_source:" src="https://cdn.jsdelivr.net/npm/@hackmd/emojify.js@2.1.0/dist/images/basic/information_source.png"><span> Since we in this exercise are most interested in exploring the pathways and comparing the abundances between the sample groups, we will only continue with the pathway abundance file.</span></p>
</div><div class="alert alert-info">
<p><img class="emoji" alt=":information_source:" src="https://cdn.jsdelivr.net/npm/@hackmd/emojify.js@2.1.0/dist/images/basic/information_source.png"><span> Activate the </span><strong><span>Humann2 Conda</span></strong><span> environment</span></p>
</div><p><span style="color:blue"><span>III] Go one directory level up (</span><kbd><span>practical/8/humann2</span></kbd><span>), and join the pathway abundance output files from the </span><strong><span>HUMAnN2</span></strong><span> runs from all samples:</span></span></p><pre><code>humann2_join_tables -i all --file_name pathabundance --output all_pathabundance.tsv

# -i, the directory of table
# -o, the table to write
# --file_name, only join tables with this string included in the file name

OUTPUT: all_pathabundance.tsv
</code></pre><p><span style="color:blue"><span>IV] Make a copy of the </span><code>all_pathabundance.tsv</code><span> and name it </span><code>all_pathabundance_filter.tsv</code><span> and open this file in a text editor.</span></span></p><div class="alert alert-info">
<p><img class="emoji" alt=":information_source:" src="https://cdn.jsdelivr.net/npm/@hackmd/emojify.js@2.1.0/dist/images/basic/information_source.png"><span> </span><strong><span>HUMAnN2</span></strong><span> sorts the pathways alphabetically, and stratify by taxa (organise by taxa). The first line is the header line and each sample is given in the following columns.</span></p>
<br>
<center><img src="https://i.imgur.com/D6QDeT7.png" alt="drawing" width="800"></center>
<br>
<p><span>All UNINTEGRATED and UNCLASSIFIED reads are calculated. UNINTEGRATED is a proxy for read mass that was mapped to genes but not incorporated into pathways, while UNCLASSIFIED are reads that were not assigned to anything.</span></p>
<br>
<center><img src="https://i.imgur.com/Yml4Jyz.png" alt="drawing" width="800"></center>
</div><p><span style="color:blue"><span>V] Because these numbers UNINTEGRATED and UNCLASSIFIED reads are relatively large they will disturb the downstream analysis, so you should remove these rows. Remove the lines containing </span><strong><span>UNINTEGRATED</span></strong><span> and </span><strong><span>UNCLASSIFIED</span></strong><span> (approximately 20 lines).</span></span></p><p><span style="color:blue"><span>VI] Change the sample names from </span><code>Sample1_pair_Abundance</code><span> to </span><code>Sample1</code><span>.</span></span></p><p><span>The </span><code>all_pathabundance_filter.tsv</code><span> file should have a similar start like this:</span></p><pre><code># Pathway	Sample1	Sample2	Sample3	Sample4	Sample5	Sample6	Sample7	Sample8
1CMET2-PWY: N10-formyl-tetrahydrofolate biosynthesis	216.8960633035	66.5325011755	532.2173413937	62.9174184708	256.7945202492	166.7137540102	457.0373778173	384.6336320725
1CMET2-PWY: N10-formyl-tetrahydrofolate biosynthesis|g__Bacteroides.s__Bacteroides_fragilis	0	0	0	0	0	0	144.9809949804	0
...
</code></pre><p><span style="color:blue"><span>VII] Transform the absolute abundance to relative abundance - so that the counts for each sample sum to 100:</span></span></p><pre><code>humann2_renorm_table --input all_pathabundance_filter.tsv --output all_pathabundance_filter_relab.tsv --units relab

OUTPUT: all_pathabundance_filter_relab.tsv
</code></pre><p><span>The </span><code>all_pathabundance_filter_relab.tsv</code><span> file should have a similar start like this:</span></p><pre><code># Pathway	Sample1	Sample2	Sample3	Sample4	Sample5	Sample6	Sample7	Sample8
1CMET2-PWY: N10-formyl-tetrahydrofolate biosynthesis	0.000101097	7.7778e-05	0.000245161	4.87381e-05	8.47476e-05	0.000140645	0.00027871	0.000163597
1CMET2-PWY: N10-formyl-tetrahydrofolate biosynthesis|g__Bacteroides.s__Bacteroides_fragilis	0	0	0	0	0	0	8.84121e-05	0
</code></pre><p><span>❓The pathway "1CMET2-PWY: N10-formyl-tetrahydrofolate biosynthesis is present in all samples. What species are contributing to this pathway from Sample 5?</span></p><details>
<summary>Solution - Click to expand 
</summary>
<div style="background-color:#c9cace">
*Escherichia coli*
</div></details><br><div class="alert alert-info">
<p><img class="emoji" alt=":information_source:" src="https://cdn.jsdelivr.net/npm/@hackmd/emojify.js@2.1.0/dist/images/basic/information_source.png"><span> </span><strong><span>HUMAnN2</span></strong><span> also provides an in-house way to test for statistical associations between metadata and the features with the command </span><code>humann2_associate</code><span>. This will perform a Kruskal-Wallis H-test between each feature’s community total and the metadata that is chosen to test for.</span></p>
<p><strong><span>HUMAnN2</span></strong><span> doesn’t generate or attach metadata - this is something you’d have to do on your own…</span></p>
<p><span>In this case we will test for associations of the treatment (Probiotics vs untreated). In a previous exercise you were given all the available metadata of these samples including the treatment. You can have a look in this file (maybe you would like to test for other associations) or just add the second line from the example below.</span></p>
</div><p><span style="color:blue"><span>VIII] Make a copy of the </span><code>all_pathabundance_filter_relab.tsv</code><span> file and name it </span><code>all_pathabundance_filter_relab_META.tsv</code><span>. Add a row below the header line with metadata for Treatment (if you prefer you can use Yes/No like in the </span><code>metadata.txt</code><span>).</span></span></p><p><span>The </span><code>all_pathabundance_filter_relab_META.tsv</code><span> file should have a similar start like this:</span></p><pre><code># Pathway	Sample1	Sample2	Sample3	Sample4	Sample5	Sample6	Sample7	Sample8
Treatment	Treated	Treated	Treated	Treated	Unreated	Unreated	Unreated	Unreated
1CMET2-PWY: N10-formyl-tetrahydrofolate biosynthesis	0.000101097	7.7778e-05	0.000245161	4.87381e-05	8.47476e-05	0.000140645	0.00027871	0.000163597
1CMET2-PWY: N10-formyl-tetrahydrofolate biosynthesis|g__Bacteroides.s__Bacteroides_fragilis	0	0	0	0	0	0	8.84121e-05	0
</code></pre><p><span style="color:blue"><span>XI] Associate pathway totals with Treatment:</span></span></p><pre><code>humann2_associate --input all_pathabundance_filter_relab_META.tsv --last-metadatum Treatment --focal-metadatum Treatment --focal-type categorical --output stats_treatment
</code></pre><p><span style="color:blue"><span>X] Open the </span><code>stats_treatment</code><span>. The summary lists treatment means and ranking by statistical significance.</span></span></p><p><span>❓How many pathways seems to be differentially enriched between the samples from patients treated with probiotics and the four untreated patients?</span></p><div class="alert alert-danger">
<p><span>Fill in the answer in the shared </span><a href="https://docs.google.com/spreadsheets/d/1JQmH5j6ygClMxUfMn654lhGKU560LoSAXW61DA03Zg8/edit?usp=sharing" target="_blank" rel="noopener"><span>Google docs table</span></a></p>
</div><p><span>❓What is the mean fractional abundance for “</span><strong><span>ARG+POLYAMINE-SYN: superpathway of arginine and polyamine biosynthesis</span></strong><span>” in the four samples from patients treated with probiotics and the four untreated patients?</span></p><details>
<summary>Solution - Click to expand 
</summary>
<div style="background-color:#c9cace">
Unreated:0.001444|Treated:0, So the samples from the treated patients does not have this pathway
</div></details><br><div class="alert alert-info">
<p><img class="emoji" alt=":information_source:" src="https://cdn.jsdelivr.net/npm/@hackmd/emojify.js@2.1.0/dist/images/basic/information_source.png"><span> This implies that the samples from the treated patients does not have this pathway.</span></p>
</div><p><span style="color:blue"><span>XI] Now that we have identified this enrichment in the untreated samples, explore the stratifications (taxa) of this pathway from the </span><code>all_pathabundance_filter_relab_META.tsv</code><span> file.</span></span></p><p><span>Alternatively use </span><strong><span>grep</span></strong><span>:</span></p><pre><code>grep 'ARG+POLYAMINE-SYN' all_pathabundance_filter_relab_META.tsv | less -S
</code></pre><p><span>❓Are there any major contributing species that seem reasonable for the enrichment of this pathway?</span></p><details>
<summary>Solution - Click to expand 
</summary>
<div style="background-color:#c9cace">
*Escherichia coli*
</div></details><br><p><span>❓How does this fit with the taxonomic analysis you did in Module VII? Hint: you could check by opening MEGAN…</span></p><details>
<summary>Solution - Click to expand 
</summary>
<div style="background-color:#c9cace">
*Escherichia coli* is highly abundant in the untreated samples, and very low abundant in the treated samples
</div></details><br><p><span>❓The “</span><strong><span>ARGSYN-PWY: L-arginine biosynthesis I (via L-ornithine)</span></strong><span>” pathway seems to be enriched in the samples from the patients treated with probiotics. Which species are contributing for the enrichment of this pathway?</span></p><details>
<summary>Solution - Click to expand 
</summary>
Bifidobacterium longum and Staphylococcus epidermidis
</details><br><div class="alert alert-info">
<p><img class="emoji" alt=":information_source:" src="https://cdn.jsdelivr.net/npm/@hackmd/emojify.js@2.1.0/dist/images/basic/information_source.png"><span> </span><strong><span>HUMAnN2</span></strong><span> includes a utility </span><code>humann2_barplot</code><span> to assist with visualising stratified outputs and produces plots of stratified HUMAnN2 features.</span></p>
</div><p><span style="color:blue"><span>XII] Make a plot of the significantly differentially abundant pathway we identified above.</span></span></p><pre><code>humann2_barplot --input all_pathabundance_filter_relab_META.tsv --focal-feature ARGSYN-PWY --focal-metadatum Treatment --last-metadatum Treatment --output ARGSYN-PWY_treatment1.png
</code></pre><p><span style="color:blue"><span>XIII] Open the plot:</span></span></p><center><img src="https://i.imgur.com/qw1OV8a.png" alt="drawing" width="700"></center><br><p><span>❓Which species seem to harbour this pathway?</span></p><details>
<summary>Solution - Click to expand 
</summary>
<div style="background-color:#c9cace">
Bifidobacterium longum, Staphylococcus epidermidis and Escherichia coli
</div></details><br><p><span style="color:blue"><span>XIV] Make a similar plot as above, but add the sort flag.</span></span></p><pre><code>humann2_barplot --sort sum --input all_pathabundance_filter_relab_META.tsv --focal-feature ARGSYN-PWY --focal-metadatum Treatment --last-metadatum Treatment --output ARGSYN-PWY_treatment2.png
</code></pre><p><span>❓What does the sort function do with the plot?</span></p><details>
<summary>Solution - Click to expand 
</summary>
<div style="background-color:#c9cace">
Order samples from highest to lowest abundance for this pathway
</div></details><br><div class="alert alert-warning">
<h4 id="-Optional-exercise1" data-id="-Optional-exercise"><a class="anchor hidden-xs" href="#-Optional-exercise1" title="-Optional-exercise1"><span class="octicon octicon-link"></span></a><img class="emoji" alt=":white_check_mark:" src="https://cdn.jsdelivr.net/npm/@hackmd/emojify.js@2.1.0/dist/images/basic/white_check_mark.png"><span> Optional exercise</span></h4>
<p><span style="color:blue"><span>Examine the “</span><strong><span>BRANCHED-CHAIN-AA-SYN-PWY: superpathway of branched amino acid biosynthesis</span></strong><span>” pathway which is enriched in the samples from the patients treated with probiotics.</span></span></p>
<br>
<p><span>❓Which species are contributing for the enrichment of this pathway?</span></p>
<details>
<summary>Solution - Click to expand 
</summary>
<div style="background-color:#c9cace">
*Bifidobacterium longum*, *Staphylococcus epidermidis* and *Streptococcus mitis oralis_pneumoniae*
</div></details>
<br>
<p><span style="color:blue"><span>Make a plot of the “</span><strong><span>BRANCHED-CHAIN-AA-SYN-PWY: superpathway of branched amino acid biosynthesis</span></strong><span>” pathway.</span></span></p>
<h4 id="It-is-possible-to-digg-deeper-into-each-pathway-and-try-to-correlate-this-with-the-results-from-SUPER-FOCUS-But-unfortunately-we-do-not-have-time-to-do-this-in-this-course😩" data-id="It-is-possible-to-digg-deeper-into-each-pathway-and-try-to-correlate-this-with-the-results-from-SUPER-FOCUS-But-unfortunately-we-do-not-have-time-to-do-this-in-this-course😩"><a class="anchor hidden-xs" href="#It-is-possible-to-digg-deeper-into-each-pathway-and-try-to-correlate-this-with-the-results-from-SUPER-FOCUS-But-unfortunately-we-do-not-have-time-to-do-this-in-this-course😩" title="It-is-possible-to-digg-deeper-into-each-pathway-and-try-to-correlate-this-with-the-results-from-SUPER-FOCUS-But-unfortunately-we-do-not-have-time-to-do-this-in-this-course😩"><span class="octicon octicon-link"></span></a><span>It is possible to digg deeper into each pathway and try to correlate this with the results from SUPER-FOCUS. But unfortunately we do not have time to do this in this course😩</span></h4>
</div><summary><span style="font-weight:bold">Progress tracker</span></summary><div style="padding:1px;background:#CCC;">
 <div style="width:100%;background:#291;text-align:center;">
   <span style="color:white"><span>Complete</span>
 </span></div>
</div><hr><h2 id="That-was-the-end-of-the-this-practical---Good-job-👍" data-id="That-was-the-end-of-the-this-practical---Good-job-👍"><a class="anchor hidden-xs" href="#That-was-the-end-of-the-this-practical---Good-job-👍" title="That-was-the-end-of-the-this-practical---Good-job-👍"><span class="octicon octicon-link"></span></a><span>That was the end of the this practical - Good job 👍</span></h2><hr></div>
    <div class="ui-toc dropup unselectable hidden-print" style="display:none;">
        <div class="pull-right dropdown">
            <a id="tocLabel" class="ui-toc-label btn btn-default" data-toggle="dropdown" href="#" role="button" aria-haspopup="true" aria-expanded="false" title="Table of content">
                <i class="fa fa-bars"></i>
            </a>
            <ul id="ui-toc" class="ui-toc-dropdown dropdown-menu" aria-labelledby="tocLabel">
                <div class="toc"><ul class="nav">
<li><a href="#8-Functional-classification" title="8. Functional classification">8. Functional classification</a><ul class="nav">
<li><a href="#1-Functional-annotation-of-reads" title="1. Functional annotation of reads">1. Functional annotation of reads</a></li>
<li><a href="#2-Visualisation-of-functional-profiles" title="2. Visualisation of functional profiles">2. Visualisation of functional profiles</a></li>
<li><a href="#3-Pathway-analysis-using-HUMMan2" title="3. Pathway analysis using HUMMan2">3. Pathway analysis using HUMMan2</a></li>
<li><a href="#That-was-the-end-of-the-this-practical---Good-job-👍" title="That was the end of the this practical - Good job 👍">That was the end of the this practical - Good job 👍</a></li>
</ul>
</li>
</ul>
</div><div class="toc-menu"><a class="expand-toggle" href="#">Expand all</a><a class="back-to-top" href="#">Back to top</a><a class="go-to-bottom" href="#">Go to bottom</a></div>
            </ul>
        </div>
    </div>
    <div id="ui-toc-affix" class="ui-affix-toc ui-toc-dropdown unselectable hidden-print" data-spy="affix" style="top:17px;display:none;" null null>
        <div class="toc"><ul class="nav">
<li><a href="#8-Functional-classification" title="8. Functional classification">8. Functional classification</a><ul class="nav">
<li><a href="#1-Functional-annotation-of-reads" title="1. Functional annotation of reads">1. Functional annotation of reads</a></li>
<li><a href="#2-Visualisation-of-functional-profiles" title="2. Visualisation of functional profiles">2. Visualisation of functional profiles</a></li>
<li><a href="#3-Pathway-analysis-using-HUMMan2" title="3. Pathway analysis using HUMMan2">3. Pathway analysis using HUMMan2</a></li>
<li><a href="#That-was-the-end-of-the-this-practical---Good-job-👍" title="That was the end of the this practical - Good job 👍">That was the end of the this practical - Good job 👍</a></li>
</ul>
</li>
</ul>
</div><div class="toc-menu"><a class="expand-toggle" href="#">Expand all</a><a class="back-to-top" href="#">Back to top</a><a class="go-to-bottom" href="#">Go to bottom</a></div>
    </div>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.1.1/jquery.min.js" integrity="sha256-hVVnYaiADRTO2PzUGmuLJr8BLUSjGIZsDYGmIJLv2b8=" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous" defer></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/gist-embed/2.6.0/gist-embed.min.js" integrity="sha256-KyF2D6xPIJUW5sUDSs93vWyZm+1RzIpKCexxElmxl8g=" crossorigin="anonymous" defer></script>
    <script>
        var markdown = $(".markdown-body");
        //smooth all hash trigger scrolling
        function smoothHashScroll() {
            var hashElements = $("a[href^='#']").toArray();
            for (var i = 0; i < hashElements.length; i++) {
                var element = hashElements[i];
                var $element = $(element);
                var hash = element.hash;
                if (hash) {
                    $element.on('click', function (e) {
                        // store hash
                        var hash = this.hash;
                        if ($(hash).length <= 0) return;
                        // prevent default anchor click behavior
                        e.preventDefault();
                        // animate
                        $('body, html').stop(true, true).animate({
                            scrollTop: $(hash).offset().top
                        }, 100, "linear", function () {
                            // when done, add hash to url
                            // (default click behaviour)
                            window.location.hash = hash;
                        });
                    });
                }
            }
        }

        smoothHashScroll();
        var toc = $('.ui-toc');
        var tocAffix = $('.ui-affix-toc');
        var tocDropdown = $('.ui-toc-dropdown');
        //toc
        tocDropdown.click(function (e) {
            e.stopPropagation();
        });

        var enoughForAffixToc = true;

        function generateScrollspy() {
            $(document.body).scrollspy({
                target: ''
            });
            $(document.body).scrollspy('refresh');
            if (enoughForAffixToc) {
                toc.hide();
                tocAffix.show();
            } else {
                tocAffix.hide();
                toc.show();
            }
            $(document.body).scroll();
        }

        function windowResize() {
            //toc right
            var paddingRight = parseFloat(markdown.css('padding-right'));
            var right = ($(window).width() - (markdown.offset().left + markdown.outerWidth() - paddingRight));
            toc.css('right', right + 'px');
            //affix toc left
            var newbool;
            var rightMargin = (markdown.parent().outerWidth() - markdown.outerWidth()) / 2;
            //for ipad or wider device
            if (rightMargin >= 133) {
                newbool = true;
                var affixLeftMargin = (tocAffix.outerWidth() - tocAffix.width()) / 2;
                var left = markdown.offset().left + markdown.outerWidth() - affixLeftMargin;
                tocAffix.css('left', left + 'px');
            } else {
                newbool = false;
            }
            if (newbool != enoughForAffixToc) {
                enoughForAffixToc = newbool;
                generateScrollspy();
            }
        }
        $(window).resize(function () {
            windowResize();
        });
        $(document).ready(function () {
            windowResize();
            generateScrollspy();
        });

        //remove hash
        function removeHash() {
            window.location.hash = '';
        }

        var backtotop = $('.back-to-top');
        var gotobottom = $('.go-to-bottom');

        backtotop.click(function (e) {
            e.preventDefault();
            e.stopPropagation();
            if (scrollToTop)
                scrollToTop();
            removeHash();
        });
        gotobottom.click(function (e) {
            e.preventDefault();
            e.stopPropagation();
            if (scrollToBottom)
                scrollToBottom();
            removeHash();
        });

        var toggle = $('.expand-toggle');
        var tocExpand = false;

        checkExpandToggle();
        toggle.click(function (e) {
            e.preventDefault();
            e.stopPropagation();
            tocExpand = !tocExpand;
            checkExpandToggle();
        })

        function checkExpandToggle () {
            var toc = $('.ui-toc-dropdown .toc');
            var toggle = $('.expand-toggle');
            if (!tocExpand) {
                toc.removeClass('expand');
                toggle.text('Expand all');
            } else {
                toc.addClass('expand');
                toggle.text('Collapse all');
            }
        }

        function scrollToTop() {
            $('body, html').stop(true, true).animate({
                scrollTop: 0
            }, 100, "linear");
        }

        function scrollToBottom() {
            $('body, html').stop(true, true).animate({
                scrollTop: $(document.body)[0].scrollHeight
            }, 100, "linear");
        }
    </script>
</body>

</html>
